broker/src/main/java/io/camunda/zeebe/broker/system/restore/PartitionRestoreService.java

```diff
@@ -0,0 +1,250 @@
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Restores a backup from the given {@link io.camunda.zeebe.backup.api.BackupStore}. */
+public class PartitionRestoreService {
+  private static final Logger LOG = LoggerFactory.getLogger(PartitionRestoreService.class);
+  final BackupStore backupStore;
+  final int partitionId;
+
+  // All members of the cluster. A backup could have been taken by any broker. So we have to iterate
+  // over all of them to find a valid backup for this partition with the given id.
+  final Set<Integer> brokerIds;
+  final Path rootDirectory;
+  private final RaftPartition partition;
+  private final int localBrokerId;
+
+  public PartitionRestoreService(
+      final BackupStore backupStore,
+      final RaftPartition partition,
+      final Set<Integer> brokerIds,
+      final int localNodeId) {
+    this.backupStore = backupStore;
+    partitionId = partition.id().id();
+    rootDirectory = partition.dataDirectory().toPath();
+    this.partition = partition;
+    this.brokerIds = brokerIds;
+    localBrokerId = localNodeId;
+  }
+
+  /**
+   * Downloads backup from the backup file, restore it to the partition's data directory. After
+   * restoring, it truncates the journal to the checkpointPosition so that the last record in the
+   * journal will be the checkpoint record at checkpointPosition.
+   *
+   * @param backupId id of the backup to restore from
+   * @return the descriptor of the backup it restored
+   */
+  public CompletableFuture<BackupDescriptor> restore(final long backupId) {
+    return getTargetDirectory(backupId)
+        .thenCompose(targetDirectory -> download(backupId, targetDirectory))
+        .thenApply(this::moveFilesToDataDirectory)
+        .thenApply(
+            backup -> {
+              resetLogToCheckpointPosition(backup.descriptor().checkpointPosition(), rootDirectory);
+              return backup.descriptor();
+            })
+        .toCompletableFuture();
+
+    // TODO: As an additional consistency check:
+    // - Validate journal.firstIndex <= snapshotIndex + 1
+    // - Verify journal.lastEntry.asqn == checkpointPosition
+  }
+
+  private CompletionStage<Path> getTargetDirectory(final long backupId) {
+    try {
+      if (!isEmpty(rootDirectory)) {
+        LOG.error(
+            "Partition's data directory {} is not empty. Aborting restore to avoid overwriting data. Please restart with a clean directory.",
+            rootDirectory);
+        CompletableFuture.failedFuture(new DirectoryNotEmptyException(rootDirectory.toString()));
+      }
+
+      // First download the contents to a temporary directory and then move it to the correct
+      // locations.
+      final var tempTargetDirectory = rootDirectory.resolve("restoring-" + backupId);
+      ensureDirectoryExists(tempTargetDirectory);
```

Comment:
No. We are not cleaning up later. We are relying on the user to do it. For now, at least the assumption is that this is done manually. So it might be ok. But it is a good point, we can probably do it.

Reply:
I thought about it a bit more. It would be better to do it in the coordinator that restores all partition. If any partition fails to restore, then we can clean up the whole data directory of the broker, not just one partition. I will take care of it in the follow up PR which adds the coordinator.